File: docksmith/layers.py
# docksmith/layers.py
# ============================================================
#  PROTHAM — File 2: Layer Creation & Storage
# ============================================================

import os
import io
import tarfile
import hashlib
import shutil
import tempfile

LAYERS_DIR = os.path.expanduser("~/.docksmith/layers")


def ensure_layers_dir():
    os.makedirs(LAYERS_DIR, exist_ok=True)


def sha256_of_bytes(data: bytes) -> str:
    return "sha256:" + hashlib.sha256(data).hexdigest()


def sha256_of_file(filepath: str) -> str:
    h = hashlib.sha256()
    with open(filepath, "rb") as f:
        for chunk in iter(lambda: f.read(65536), b""):
            h.update(chunk)
    return h.hexdigest()


def create_delta_tar(base_dir: str, paths_to_include: list) -> bytes:
    """
    Creates a reproducible tar archive.
    CRITICAL: sorted entries + zeroed timestamps.
    """
    buf = io.BytesIO()
    sorted_paths = sorted(set(paths_to_include))

    with tarfile.open(fileobj=buf, mode="w:") as tar:
        for abs_path in sorted_paths:
            if not os.path.exists(abs_path):
                continue
            rel_path = os.path.relpath(abs_path, base_dir)
            tarinfo  = tar.gettarinfo(name=abs_path, arcname=rel_path)
            tarinfo.mtime  = 0
            tarinfo.uid    = 0
            tarinfo.gid    = 0
            tarinfo.uname  = ""
            tarinfo.gname  = ""
            if tarinfo.isreg():
                with open(abs_path, "rb") as f:
                    tar.addfile(tarinfo, f)
            else:
                tar.addfile(tarinfo)

    return buf.getvalue()


def collect_all_paths(directory: str) -> list:
    """
    Return all paths under directory in sorted order.
    Uses os.scandir so directory-symlinks (e.g. bin -> usr/bin on Debian 12)
    are included as entries rather than being silently skipped by os.walk.
    """
    all_paths = []

    def _walk(path: str):
        try:
            entries = sorted(os.scandir(path), key=lambda e: e.name)
        except PermissionError:
            return
        for entry in entries:
            all_paths.append(entry.path)
            # Recurse into real directories only; do NOT follow symlinks
            if entry.is_dir(follow_symlinks=False):
                _walk(entry.path)

    _walk(directory)
    return all_paths


def store_layer(tar_bytes: bytes) -> str:
    ensure_layers_dir()
    digest   = sha256_of_bytes(tar_bytes)
    hex_hash = digest.replace("sha256:", "")
    layer_path = os.path.join(LAYERS_DIR, hex_hash)

    if not os.path.exists(layer_path):
        tmp_path = layer_path + ".tmp"
        try:
            with open(tmp_path, "wb") as f:
                f.write(tar_bytes)
            os.rename(tmp_path, layer_path)
        except Exception:
            if os.path.exists(tmp_path):
                os.remove(tmp_path)
            raise

    return digest


def layer_exists(layer_digest: str) -> bool:
    hex_hash = layer_digest.replace("sha256:", "")
    return os.path.exists(os.path.join(LAYERS_DIR, hex_hash))


def get_layer_size(layer_digest: str) -> int:
    hex_hash   = layer_digest.replace("sha256:", "")
    layer_path = os.path.join(LAYERS_DIR, hex_hash)
    if not os.path.exists(layer_path):
        raise FileNotFoundError(f"[LAYER ERROR] Layer not found: {layer_digest}")
    return os.path.getsize(layer_path)


def _ensure_usrmerge(rootfs: str):
    """
    Debian 12+ uses 'usrmerge': /bin, /sbin, /lib, /lib64 are symlinks to
    their /usr/ counterparts.  Our re-tar code uses os.walk which silently
    drops directory-symlinks, so we restore them here after extraction.
    """
    usrmerge = {
        "bin":    "usr/bin",
        "sbin":   "usr/sbin",
        "lib":    "usr/lib",
        "lib32":  "usr/lib32",
        "lib64":  "usr/lib64",
        "libx32": "usr/libx32",
    }
    for link_name, target in usrmerge.items():
        link_path   = os.path.join(rootfs, link_name)
        target_path = os.path.join(rootfs, target)
        # Only create the symlink when the link is absent but the target exists
        if not os.path.exists(link_path) and not os.path.islink(link_path):
            if os.path.isdir(target_path):
                os.symlink(target, link_path)


def extract_layer(layer_digest: str, target_dir: str):
    hex_hash   = layer_digest.replace("sha256:", "")
    layer_path = os.path.join(LAYERS_DIR, hex_hash)
    if not os.path.exists(layer_path):
        raise FileNotFoundError(f"[RUNTIME ERROR] Layer not found: {layer_digest}")
    with tarfile.open(layer_path, "r:") as tar:
        try:
            # Use 'fully_trusted' — container rootfs layers contain absolute symlinks
            # (e.g. /etc/alternatives/awk → /usr/bin/mawk) which are expected.
            tar.extractall(path=target_dir, filter="fully_trusted")
        except TypeError:
            # Python < 3.12 doesn't support 'filter'
            tar.extractall(path=target_dir)
    # Restore Debian usrmerge symlinks that may have been stripped by re-tarring
    _ensure_usrmerge(target_dir)


def delete_layer(layer_digest: str):
    hex_hash   = layer_digest.replace("sha256:", "")
    layer_path = os.path.join(LAYERS_DIR, hex_hash)
    if os.path.exists(layer_path):
        os.remove(layer_path)


def hash_copy_sources(src_pattern: str, context_dir: str) -> list:
    import glob as glob_module

    if src_pattern == ".":
        matched = []
        for root, dirs, files in os.walk(context_dir):
            dirs.sort()
            for f in sorted(files):
                matched.append(os.path.join(root, f))
    else:
        full_pattern = os.path.join(context_dir, src_pattern)
        matched = sorted(glob_module.glob(full_pattern, recursive=True))

    result = []
    for abs_path in sorted(matched):
        if os.path.isfile(abs_path):
            rel = os.path.relpath(abs_path, context_dir)
            h   = sha256_of_file(abs_path)
            result.append(f"{rel}:{h}")
    return sorted(result)


def snapshot_filesystem(directory: str) -> dict:
    snapshot = {}
    for root, dirs, files in os.walk(directory):
        dirs.sort()
        for fname in sorted(files):
            abs_path = os.path.join(root, fname)
            rel_path = os.path.relpath(abs_path, directory)
            try:
                snapshot[rel_path] = sha256_of_file(abs_path)
            except (PermissionError, OSError):
                pass
    return snapshot


def compute_delta_paths(before: dict, after: dict, base_dir: str) -> list:
    changed = []
    for rel_path, new_hash in after.items():
        if rel_path not in before or before[rel_path] != new_hash:
            changed.append(os.path.join(base_dir, rel_path))
    return sorted(changed)

File: internal/layers/layers.go
package layers

import (
	"archive/tar"
	"bytes"
	"crypto/sha256"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"time"

	"docksmith/internal/paths"
)

// SHA256OfBytes returns "sha256:<hexdigest>" for the given bytes.
func SHA256OfBytes(data []byte) string {
	h := sha256.Sum256(data)
	return fmt.Sprintf("sha256:%x", h)
}

// SHA256OfFile returns the hex digest (no prefix) of a file's contents.
func SHA256OfFile(path string) (string, error) {
	f, err := os.Open(path)
	if err != nil {
		return "", err
	}
	defer f.Close()
	h := sha256.New()
	if _, err := io.Copy(h, f); err != nil {
		return "", err
	}
	return fmt.Sprintf("%x", h.Sum(nil)), nil
}

// CollectAllPaths returns all file and directory paths under dir, sorted lexicographically.
func CollectAllPaths(dir string) ([]string, error) {
	var paths []string
	err := filepath.Walk(dir, func(p string, info os.FileInfo, err error) error {
		if err != nil {
			return err
		}
		if p == dir {
			return nil
		}
		paths = append(paths, p)
		return nil
	})
	sort.Strings(paths)
	return paths, err
}

// CreateDeltaTar creates a reproducible tar archive (zeroed timestamps, sorted entries).
// Returns raw tar bytes.
func CreateDeltaTar(baseDir string, absPaths []string) ([]byte, error) {
	// Deduplicate and sort
	seen := make(map[string]bool)
	var unique []string
	for _, p := range absPaths {
		if !seen[p] {
			seen[p] = true
			unique = append(unique, p)
		}
	}
	sort.Strings(unique)

	var buf bytes.Buffer
	tw := tar.NewWriter(&buf)

	for _, absPath := range unique {
		info, err := os.Lstat(absPath)
		if err != nil {
			continue // skip if gone
		}
		relPath, err := filepath.Rel(baseDir, absPath)
		if err != nil {
			return nil, err
		}
		// Normalise to forward slashes
		relPath = filepath.ToSlash(relPath)

		hdr := &tar.Header{
			Name:     relPath,
			Mode:     int64(info.Mode()),
			ModTime:  time.Time{},
			Uid:      0,
			Gid:      0,
			Uname:    "",
			Gname:    "",
			Typeflag: tar.TypeReg,
		}

		if info.IsDir() {
			hdr.Typeflag = tar.TypeDir
			hdr.Name = relPath + "/"
			if err := tw.WriteHeader(hdr); err != nil {
				return nil, err
			}
			continue
		}

		if info.Mode()&os.ModeSymlink != 0 {
			target, err := os.Readlink(absPath)
			if err != nil {
				return nil, err
			}
			hdr.Typeflag = tar.TypeSymlink
			hdr.Linkname = target
			hdr.Size = 0
			if err := tw.WriteHeader(hdr); err != nil {
				return nil, err
			}
			continue
		}

		hdr.Size = info.Size()
		if err := tw.WriteHeader(hdr); err != nil {
			return nil, err
		}
		f, err := os.Open(absPath)
		if err != nil {
			return nil, err
		}
		if _, err := io.Copy(tw, f); err != nil {
			f.Close()
			return nil, err
		}
		f.Close()
	}

	if err := tw.Close(); err != nil {
		return nil, err
	}
	return buf.Bytes(), nil
}

// StoreLayer writes tar bytes to ~/.docksmith/layers/<hexhash>.
// Returns "sha256:<hexhash>". Skips write if already present.
func StoreLayer(tarBytes []byte) (string, error) {
	if err := paths.EnsureStateDirs(); err != nil {
		return "", err
	}
	layersDir, err := paths.LayersDir()
	if err != nil {
		return "", err
	}

	digest := SHA256OfBytes(tarBytes)
	hexHash := strings.TrimPrefix(digest, "sha256:")
	layerPath := filepath.Join(layersDir, hexHash)

	if _, err := os.Stat(layerPath); err == nil {
		return digest, nil // already exists
	}

	tmpPath := layerPath + ".tmp"
	if err := os.WriteFile(tmpPath, tarBytes, 0644); err != nil {
		return "", err
	}
	if err := os.Rename(tmpPath, layerPath); err != nil {
		os.Remove(tmpPath)
		return "", err
	}
	return digest, nil
}

// LayerExists reports whether a layer file is on disk.
func LayerExists(digest string) bool {
	layersDir, err := paths.LayersDir()
	if err != nil {
		return false
	}
	hexHash := strings.TrimPrefix(digest, "sha256:")
	_, err = os.Stat(filepath.Join(layersDir, hexHash))
	return err == nil
}

// GetLayerSize returns the byte size of a stored layer.
func GetLayerSize(digest string) (int64, error) {
	layersDir, err := paths.LayersDir()
	if err != nil {
		return 0, err
	}
	hexHash := strings.TrimPrefix(digest, "sha256:")
	info, err := os.Stat(filepath.Join(layersDir, hexHash))
	if err != nil {
		return 0, fmt.Errorf("[LAYER ERROR] layer not found: %s", digest)
	}
	return info.Size(), nil
}

// ExtractLayer extracts a stored layer tar into targetDir.
func ExtractLayer(digest, targetDir string) error {
	layersDir, err := paths.LayersDir()
	if err != nil {
		return err
	}
	hexHash := strings.TrimPrefix(digest, "sha256:")
	layerPath := filepath.Join(layersDir, hexHash)

	f, err := os.Open(layerPath)
	if err != nil {
		return fmt.Errorf("[RUNTIME ERROR] layer not found: %s — try rebuilding with --no-cache", digest)
	}
	defer f.Close()

	tr := tar.NewReader(f)
	for {
		hdr, err := tr.Next()
		if err == io.EOF {
			break
		}
		if err != nil {
			return err
		}

		target := filepath.Join(targetDir, filepath.FromSlash(hdr.Name))

		switch hdr.Typeflag {
		case tar.TypeDir:
			if err := os.MkdirAll(target, os.FileMode(hdr.Mode)|0755); err != nil {
				return err
			}
		case tar.TypeSymlink:
			os.Remove(target)
			if err := os.Symlink(hdr.Linkname, target); err != nil {
				return err
			}
		default:
			if err := os.MkdirAll(filepath.Dir(target), 0755); err != nil {
				return err
			}
			out, err := os.OpenFile(target, os.O_CREATE|os.O_WRONLY|os.O_TRUNC, os.FileMode(hdr.Mode))
			if err != nil {
				return err
			}
			if _, err := io.Copy(out, tr); err != nil {
				out.Close()
				return err
			}
			out.Close()
		}
	}
	return nil
}

// DeleteLayer removes a layer file from disk.
func DeleteLayer(digest string) error {
	layersDir, err := paths.LayersDir()
	if err != nil {
		return err
	}
	hexHash := strings.TrimPrefix(digest, "sha256:")
	path := filepath.Join(layersDir, hexHash)
	if err := os.Remove(path); err != nil && !os.IsNotExist(err) {
		return err
	}
	return nil
}

// HashCopySources returns sorted "relpath:hexdigest" strings for files matched by srcPattern.
func HashCopySources(srcPattern, contextDir string) ([]string, error) {
	var matched []string

	if srcPattern == "." {
		err := filepath.Walk(contextDir, func(p string, info os.FileInfo, err error) error {
			if err != nil {
				return err
			}
			if !info.IsDir() {
				matched = append(matched, p)
			}
			return nil
		})
		if err != nil {
			return nil, err
		}
	} else {
		pattern := filepath.Join(contextDir, srcPattern)
		m, err := filepath.Glob(pattern)
		if err != nil {
			return nil, err
		}
		matched = m
	}

	sort.Strings(matched)
	var result []string
	for _, absPath := range matched {
		info, err := os.Stat(absPath)
		if err != nil || info.IsDir() {
			continue
		}
		h, err := SHA256OfFile(absPath)
		if err != nil {
			return nil, err
		}
		rel, err := filepath.Rel(contextDir, absPath)
		if err != nil {
			return nil, err
		}
		result = append(result, filepath.ToSlash(rel)+":"+h)
	}
	sort.Strings(result)
	return result, nil
}

// SnapshotFilesystem returns a map of relpath → sha256hex for all files under dir.
func SnapshotFilesystem(dir string) (map[string]string, error) {
	snap := make(map[string]string)
	err := filepath.Walk(dir, func(p string, info os.FileInfo, err error) error {
		if err != nil || info.IsDir() {
			return nil
		}
		h, err := SHA256OfFile(p)
		if err != nil {
			return nil // skip unreadable files
		}
		rel, err := filepath.Rel(dir, p)
		if err != nil {
			return err
		}
		snap[filepath.ToSlash(rel)] = h
		return nil
	})
	return snap, err
}

// ComputeDeltaPaths returns absolute paths of files new or changed in after vs before.
func ComputeDeltaPaths(before, after map[string]string, baseDir string) []string {
	var changed []string
	for relPath, newHash := range after {
		if oldHash, ok := before[relPath]; !ok || oldHash != newHash {
			changed = append(changed, filepath.Join(baseDir, filepath.FromSlash(relPath)))
		}
	}
	sort.Strings(changed)
	return changed
}
